lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -2584,37 +2588,142 @@ lldb::addr_t SwiftLanguageRuntime::GetAsyncContext(RegisterContext *regctx) {
-      llvm::dwarf::DW_OP_deref,
-      llvm::dwarf::DW_OP_deref,
-  };
-  static const uint8_t g_cfa_dwarf_expression_arm64[] = {
-      llvm::dwarf::DW_OP_breg29, // DW_OP_breg29, register 29 == fp
-      0x78,                      //    sleb128 -8 (ptrsize)
-      llvm::dwarf::DW_OP_deref,
-      llvm::dwarf::DW_OP_deref,
-  };
+/// Functional wrapper to read a register as an address.
+static llvm::Expected<addr_t> ReadRegisterAsAddress(RegisterContext &regctx,
+                                                    RegisterKind regkind,
+                                                    unsigned regnum) {
+  unsigned lldb_regnum =
+      regctx.ConvertRegisterKindToRegisterNumber(regkind, regnum);
+  auto reg = regctx.ReadRegisterAsUnsigned(lldb_regnum, LLDB_INVALID_ADDRESS);
+  if (reg != LLDB_INVALID_ADDRESS)
+    return reg;
+  return llvm::createStringError(
+      "SwiftLanguageRuntime: failed to read register from regctx");
+}
+
+/// Functional wrapper to read a pointer from process memory at `addr +
+/// offset`.
+static llvm::Expected<addr_t>
+ReadPtrFromAddr(Process &process, llvm::Expected<addr_t> addr, int offset = 0) {
+  if (!addr)
+    return addr;
+  Status error;
+  addr_t ptr = process.ReadPointerFromMemory(*addr + offset, error);
+  if (ptr != LLDB_INVALID_ADDRESS)
+    return ptr;
+  return llvm::createStringError("SwiftLanguageRuntime: Failed to read ptr "
+                                 "from memory address 0x%8.8" PRIx64,
+                                 *addr + offset);
+}
+
+/// Computes the Canonical Frame Address (CFA) by converting the abstract
+/// location of UnwindPlan::Row::FAValue into a concrete address. This is a
+/// simplified version of the methods in RegisterContextUnwind, since plumbing
+/// access to those here would be challenging.
+static llvm::Expected<addr_t> GetCFA(Process &process, RegisterContext &regctx,
+                                     RegisterKind regkind,
+                                     UnwindPlan::Row::FAValue cfa_loc) {
+  using ValueType = UnwindPlan::Row::FAValue::ValueType;
+  switch (cfa_loc.GetValueType()) {
+  case ValueType::isRegisterPlusOffset: {
+    unsigned regnum = cfa_loc.GetRegisterNumber();
+    if (llvm::Expected<addr_t> regvalue =
+            ReadRegisterAsAddress(regctx, regkind, regnum))
+      return *regvalue + cfa_loc.GetOffset();
+    else
+      return regvalue;
+  }
+  case ValueType::isConstant:
+  case ValueType::isDWARFExpression:
+  case ValueType::isRaSearch:
+  case ValueType::isRegisterDereferenced:
+  case ValueType::unspecified:
+    break;
+  }
+  return llvm::createStringError(
+      "SwiftLanguageRuntime: Unsupported FA location type = %d",
+      cfa_loc.GetValueType());
+}
+
+/// Attempts to use UnwindPlans that inspect assembly to recover the entry value
+/// of the async context register. This is a simplified version of the methods
+/// in RegisterContextUnwind, since plumbing access to those here would be
+/// challenging.
```

Comment:
There are two such comments. Do we need to consider divergence? What's the likelihood of changes or fixes to `RegisterContextUnwind` that need to be mirrored here?

Reply:
In both methods where this comment is present, we are `switch`ing over enum entries (of different enumerations). If `RegisterContextUnwind` ever adds more cases to either of those, we will get compiler errors

Reply:
what about the cases that are covered? If those change in the original implementation, what does that mean for these implementations?

Reply:
The way these are interpreted (Row::FAValue isRegisterPlusOffset, Row::AbstractRegisterLocation RestoreType) are very simple, I could see new enum types being added (especially to the AbstractRegisterLocation) but the interpretation of these different restore types are so simple I don't think the correct implementation will change.  If it did, I believe the test cases Felipe is adding will catch them.